### ISSUE TITLE
FP: `map_clone` suggest clone, which changes the output type

### ISSUE THREAD

**Issue by xFrednet** [rust-lang/rust-clippy#10278]
### Summary

I used `.map(|s| s.clone())` on an iterator and got the suggestion to use `.clone()` instead. The suggestion caused an error, as the original clone was used to get an object from a reference, while `.clone()` on an iterator clones the references.

### Lint Name

[`map_clone`](https://rust-lang.github.io/rust-clippy/master/index.html#map_clone)

### Reproducer

I tried this code:

```rust
fn main() {
    let _: Vec<String> = ["123".to_string()].iter().map(|s| s.clone()).collect();
}
```

I saw this happen:

```rs
warning: you are using an explicit closure for cloning elements
 --> src/main.rs:4:26
  |
4 |     let _: Vec<String> = ["123".to_string()].iter().map(|s| s.clone()).collect();
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `["123".to_string()].iter().cloned()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
  = note: `#[warn(clippy::map_clone)]` on by default
```

When applying the suggestion, I got this warning:

```rs
error[[E0277]](https://doc.rust-lang.org/stable/error-index.html#E0277): a value of type `Vec<String>` cannot be built from an iterator over elements of type `&String`
 --> src/main.rs:6:26
  |
6 |     let _: Vec<String> = ["123".to_string()].iter().clone().collect();
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ------- required by a bound introduced by this call
  |                          |
  |                          value of type `Vec<String>` cannot be built from `std::iter::Iterator<Item=&String>`
  |
  = help: the trait `FromIterator<&String>` is not implemented for `Vec<String>`
  = help: the trait `FromIterator<T>` is implemented for `Vec<T>`
```

I expected to see this happen:

No warning should be emitted, as the output type was changed from `String` to `&String`. A potential fix could be to check if the type implements `Clone`, in which case `.clone()` on an iterator will not invoke the `Clone` implementation of references, but of the type. I can't think of a better heuristic rn

### Version

```text
rustc 1.69.0-nightly (c18a5e8a5 2023-01-25)
binary: rustc
commit-hash: c18a5e8a5b1afb0d7a582fe9ebad4c1996c90da3
commit-date: 2023-01-25
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

**Comment by xFrednet** [comment]
I'm not sure on which clone I forgot the **d**. Spelling is not my specialty :sweat_smile: :joy:.

Replacing the entire chain would require checking every usage, if it works with the reference as well. That sounds pretty complicated to me :sweat_smile:

**Comment by Alexendoo** [comment]
As in I think it would be clearer if the span was

```
warning: you are using an explicit closure for cloning elements
 --> src/main.rs:4:26
  |
4 |     let _: Vec<String> = ["123".to_string()].iter().map(|s| s.clone()).collect();
  |                                                    ^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `.cloned()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
  = note: `#[warn(clippy::map_clone)]` on by default
```

**Comment by xFrednet** [comment]
Ahhh, I was using the wrong method. Now I feel dumb... But the span should be easy to adjust in most cases
